wireless/agents/rate_manager_agents.py

```diff
@@ -162,31 +162,43 @@ def act(self, snr):
         return selected_mcs
 
 
-class OptimalAgent:
+class OptimalMaxThroughputAgent:
     """
-    This agent computes the best MCS based on the future SNR, i.e., the SNR at one timestep ahead with respect to the
-    current timestep.
-    The MCS is chosen to be the one that maximizes the total average number of received packets, hence its optimality.
+    Agent that maximixes the average throughput, using the info regarding the packet on which the action will be taken.
     """
 
-    def __init__(self, action_space, error_model, timestep, pkt_size):
-        self.action_space = action_space
-        self.error_model = error_model
-        self.timestep = timestep
-        self.pkt_size = pkt_size
+    def __init__(self, action_space, error_model):
+        self._max_mcs = action_space.n - 1
+        self._error_model = error_model
 
-    def act(self, snr):
-        rx_bits = np.zeros((self.action_space.n,))
-        for action_idx in range(self.action_space.n):
-            n_packets, last_pkt, pkt_size_list = misc.get_tx_pkt_size_list(action_idx, self.timestep, self.pkt_size)
-            pkt_psr = self.error_model.get_packet_success_rate(snr, action_idx, self.pkt_size)
-            last_pkt_psr = self.error_model.get_packet_success_rate(snr, action_idx, last_pkt)
+    def act(self, state, info):
+        """
+        Performs the optimal action given the info on the packet to be sent.
 
-            avg_rx_bits = self.pkt_size * pkt_psr * n_packets + last_pkt * last_pkt_psr
-            rx_bits[action_idx] = avg_rx_bits
+        Parameters
+        ----------
+        state : dict
+            Not used, kept to maintain the same signature for all agents.
+        info : dict
+            "next_pkt_size" : int
+                Size of the packet to be sent in [B]
+            "next_snr" : float
+                Noiseless SNR of the packet to be sent [dB]
 
-        selected_mcs = np.argmax(rx_bits)
-        return selected_mcs
+        Returns
+        -------
+        The optimal MCS to be used for the next packet.
+        """
+        if info["next_pkt_size"] is None:
+            return None
+
+        tot_tx_time = [dot11ad_constants.get_total_tx_time(info["next_pkt_size"], mcs)
+                       for mcs in range(self._max_mcs + 1)]
+        psr = [self._error_model.get_packet_success_rate(info["next_snr"], mcs, info["next_pkt_size"])
+               for mcs in range(self._max_mcs + 1)]
+        thr = [info["next_pkt_size"] / tx_time * p
```

Comment:
Maybe this for-cycle can be avoided using vector operations. Or you can combine all these operations in one unique for-cycle.

Reply:
You can surely define `tot_tx_time` and `psr` using a single for cycle

Reply:
i don't think `get_total_tx_time` and `get_packet_success_rate` support vector operations, so a single for loop would be the only alternative. does it really change much? i would need to pre-allocate the lists (my religion prohibits `append`ing whenever possible), declare the for loop, and do the calculations one by one. i think it would be even longer and more painful to read (for python-experts)

Reply:
Maybe you can squeeze all operations in a unique [__ for mcs in range()].
But, I think it's fine anyway. We can merge I guess